aggregated-dependency-score.go

```diff
@@ -60,21 +60,39 @@ type DependencyResolver interface {
 	GetDirectDependencies(ctx context.Context, p Package) ([]Package, error)
 }
 
-func (eval *trustwhorthinessEvaluator) evaluate(ctx context.Context, p Package) (float64, error) {
-	intrinsic, err := eval.intrinsic.EvaluateIntrinsicTrustworthiness(ctx, p)
+func (evaluator *trustwhorthinessEvaluator) evaluate(ctx context.Context, p Package, ancestors map[string]struct{}) (float64, error) {
+	intrinsic, err := evaluator.intrinsic.EvaluateIntrinsicTrustworthiness(ctx, p)
 	if err != nil {
 		return 0.0, fmt.Errorf("evaluating intrinsic trustworthiness of package: %w", err)
 	}
 
 	result := intrinsic
 
-	deps, err := eval.deps.GetDirectDependencies(ctx, p)
+	deps, err := evaluator.deps.GetDirectDependencies(ctx, p)
 	if err != nil {
 		return 0.0, fmt.Errorf("getting direct dependencies of package: %w", err)
 	}
 
 	for _, dep := range deps {
-		tPrimeQ, err := eval.evaluate(ctx, dep)
+		// XXX sometimes different names can refer to the same package,
+		// for instance with gopkg.in URLs;
+		// XXX should we consider the version as well?
+		if _, ok := ancestors[dep.Name]; ok {
+			// depedency cycle
+			// TODO emit a log
+			continue
```

Comment:
isn't that forbidden for packages? you cannot have cyclic dependencies

Reply:
nvm. it is cyclic:
* https://github.com/stretchr/testify/network/dependencies
* https://github.com/stretchr/objx/network/dependencies
objx and testify depend on each other

Reply:
That's what I thought until I discovered the Go compiler can be OK with that: https://github.com/DataDog/aggregated-dependency-score/issues/9#issuecomment-2556649395